### PR TITLE
KAFKA-4643: Improve test coverage of StreamsKafkaClient

### PR DESCRIPTION
**Pull request by adyach**
The commit brings improved test coverage for StreamsKafkaClientTest.java

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 1]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
         verifyCorrectTopicConfigs(streamsKafkaClient, topicConfigWithNoOverrides, Collections.singletonMap("cleanup.policy", "delete"));
     }
 
+    @Test
+    public void testCheckBrokerCompatibilityEmptyResponse() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: `shouldThrowStreamsExceptionOnEmptyBrokerCompatibilityResponse`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 2]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
...
+        try {
+            streamsKafkaClient.checkBrokerCompatibility(false);
+            fail("Response was not empty");
+        } catch (final StreamsException se) {
+            assertThat(se.getMessage(), equalTo("Empty response for client request."));
+        }
+    }
+
+    @Test
+    public void testCheckBrokerCompatibilityConsistentResponseType() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: The test should describe what it is doing, i.e., `shouldThrowStreamsExceptionWhenBrokerCompatibilityResponseInconsisent`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 3]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
...
 
+    @Test
+    public void testCheckBrokerCompatibilityEmptyResponse() {
+        kafkaClient.prepareResponse(null);
+        final StreamsKafkaClient streamsKafkaClient = createStreamsKafkaClient();
+        try {
+            streamsKafkaClient.checkBrokerCompatibility(false);
+            fail("Response was not empty");
+        } catch (final StreamsException se) {
+            assertThat(se.getMessage(), equalTo("Empty response for client request."));
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: nit: I don't think we should bother checking the string messages as it is quite brittle. The important thing is that the exception is thrown

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 4]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
...
+        try {
+            streamsKafkaClient.checkBrokerCompatibility(false);
+            fail("Consistent response type");
+        } catch (final StreamsException se) {
+            assertThat(se.getMessage(), equalTo("Inconsistent response type for API versions request. Expected ApiVersionsResponse but received org.apache.kafka.common.requests.ProduceResponse"));
+        }
+    }
+
+    @Test
+    public void testCheckBrokerCompatibilityRequiresBrokerVersion0101xOrHigher() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: as above. Name of the test can be more descriptive. `shouldRequireBrokerVersion0101OrHigherWhenEosDisabled`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 5]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
...
+        try {
+            streamsKafkaClient.checkBrokerCompatibility(false);
+            fail("Version is 0.10.1.x or higher");
+        } catch (final StreamsException se) {
+            assertThat(se.getMessage(), equalTo("Kafka Streams requires broker version 0.10.1.x or higher."));
+        }
+    }
+
+    @Test
+    public void testCheckBrokerCompatibilityRequiresBrokerVersion0110xOrHigher() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: `shouldRequireBrokerVersions0110OrHigherWhenEosEnabled`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 6]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
...
+        try {
+            streamsKafkaClient.checkBrokerCompatibility(true);
+            fail("Version is 0.11.0.x or higher");
+        } catch (final StreamsException se) {
+            assertThat(se.getMessage(), equalTo("Setting processing.guarantee=exactly_once requires broker version 0.11.0.x or higher."));
+        }
+    }
+
+    @Test
+    public void fetchMetadataEmptyResponse() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: again, naming of the test

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java [hunk 7]
```diff
@@ -124,6 +131,78 @@ public void shouldNotAllowNullTopicConfigs() throws Exception {
...
+        try {
+            streamsKafkaClient.fetchMetadata();
+            fail("Response was not empty");
+        } catch (final StreamsException se) {
+            assertThat(se.getMessage(), equalTo("Empty response for client request."));
+        }
+    }
+
+    @Test
+    public void fetchMetadataConsistentResponseType() {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClientTest.java]: and again.

**Review by dguy — COMMENTED** [review]
Thanks @adyach overall looks good. I just feel the test names need to be more descriptive. They should describe what it is that should happen. Also, i don't think we need to check the string messages in the exception as it is a bit brittle. A small change in the message shouldn't make the test fail.

**Comment by adyach** [comment]
@dguy addressed your comments

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6725/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6740/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by dguy** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6745/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6730/
Test PASSed (JDK 8 and Scala 2.12).

**Review by dguy — APPROVED** [review]
Thanks for the patch @adyach, LGTM

**Comment by dguy** [comment]
merged to trunk

**Pull request closed without merging**